### PR TITLE
Reimplement elemMatch

### PR DESCRIPTION
**Pull request by miripiruni**
For #135

/cc @zxqfox @tadatuta  @indutny

### CODE REVIEW

**Comment by zxqfox** [comment]
Here is the problem with back-compat:
```
block('b').def()(
  elemMatch(function(){ return this.elem === 'e1'; })('1'),
  elem('e2')('2')
);
// +
{ block: 'b', elem: 'e2' }
```

**Comment by miripiruni** [comment]
//cc @veged

**Comment by miripiruni** [comment]
@zxqfox Here is two [tests](https://github.com/bem/bem-xjst/pull/165/files#diff-e241c0081ba5875bf6cd84d3eb063aebR190) about priority.

File: test/bemhtml/tree-test.js [hunk 1]
```diff
@@ -175,6 +175,49 @@ describe('BEMHTML compiler/Tree', function() {
...
+        return this.elem === 'e1' || this.elem === 'e2';
+      }).tag()('b');
+    }, [
+      { block: 'b1', elem: 'e1' },
+      { block: 'b1', elem: 'e2' },
+      { block: 'b1', elem: 'e3' }
+    ], '<b class="b1__e1"></b><b class="b1__e2"></b><div class="b1__e3"></div>')
+  });
+
+  it('priority of elemMatch is higher than elem', function() {
```
Comment (zxqfox) [inline test/bemhtml/tree-test.js]: Actually, I don't get why?

File: test/bemhtml/tree-test.js [hunk 2]
```diff
@@ -175,6 +175,49 @@ describe('BEMHTML compiler/Tree', function() {
...
+    test(function() {
+      block('b1').def()(
+         elem('e1')('elem'),
```
Comment (zxqfox) [inline test/bemhtml/tree-test.js]: `e1` should be a `e` — that's why. ;-)

**Comment by arikon** [comment]
We decided to remove `elemMatch()` in 5.0 in favor of `elems('a', 'b', 'c', ...)`

**Comment by zxqfox** [comment]
Don't forget to remove or fix `elem('*')`

**Pull request closed without merging**